### ISSUE TITLE
Quantizer optimization investigation

### ISSUE THREAD

**Issue by akfish** [akfish/node-vibrant#11]
Investigate the possibility of optimizing color quantizer

- [x] Reduce copy operations. Do stuff in-place.
- [x] Investigate Android M's ColorCutQuantizer

**Comment by akfish** [comment]
Commit fe5b3f4 implemented a NoCopy quantizer, which is ~2.5x faster than baseline implementation. 

Baseline uses a nasty `while` loop to handle the downsampling and copy the pixels into a format that the upstream `quantize` library could consume. This wastes too much time. Two simple modifications did the trick:

* Move downsampling out of quantizer logic. Implement with native methods #12
* Modify `quantize` library to consume pixel data directly

```
F:\dev\js\node-vibrant [feature/Quantizer +3 ~8 -0 !]> gulp benchmark
[15:45:57] Requiring external module coffee-script/register
[15:45:58] Using gulpfile F:\dev\js\node-vibrant\gulpfile.coffee
[15:45:58] Starting 'coffee'...
[15:45:58] Finished 'coffee' after 11 ms
[15:45:58] Starting 'benchmark'...
[15:45:58] Finished 'benchmark' after 4.2 ms
[15:45:58] Running suite node-vibrant quantizer benchmark [F:\dev\js\node-vibrant\test
e]...
[15:46:05]    Quantizer: Baseline x 4.36 ops/sec ±5.24% (26 runs sampled)
[15:46:11]    Quantizer: NoCopy x 10.81 ops/sec ±2.57% (56 runs sampled)
[15:46:11] Fastest test is Quantizer: NoCopy at 2.5x faster than Quantizer: Baseline
```

**Comment by akfish** [comment]
Initial ColorCut quantizer achieved ~3.2x speed-up. 
```
F:\dev\js\node-vibrant [feature/Quantizer +3 ~8 -0 !]> gulp benchmark
[23:02:45] Requiring external module coffee-script/register
[23:02:46] Using gulpfile F:\dev\js\node-vibrant\gulpfile.coffee
[23:02:46] Starting 'coffee'...
[23:02:46] Finished 'coffee' after 14 ms
[23:02:46] Starting 'benchmark'...
[23:02:46] Finished 'benchmark' after 6.03 ms
[23:02:46] Running suite node-vibrant quantizer benchmark [F:\dev\js\node-vibrant\test\v
e]...
[23:02:54]    Quantizer: Baseline x 3.14 ops/sec ±6.86% (20 runs sampled)
[23:02:59]    Quantizer: ColorCut x 9.92 ops/sec ±3.25% (52 runs sampled)
[23:02:59] Fastest test is Quantizer: ColorCut at 3.2x faster than Quantizer: Baseline
```
However the result seems to be off. Perhaps some precision problem with the bitwise operations. Also endianness could be a problem when running on different platforms.

I'll use NoCopy quantizer for now and complete other tasks at hand. Fix it later if time permits. Otherwise ColorCut quantizer will not be shipped within the time frame of 2.0.0 milestone.

- [ ] Revisit 32bit pixel manipulation

**Comment by akfish** [comment]
~3.5x-4x speed-up with MMCQ quantizer. Used as default.
```
F:\dev\js\node-vibrant [master +0 ~1 -0]> gulp benchmark
[11:51:45] Requiring external module coffee-script/register
[11:51:48] Using gulpfile F:\dev\js\node-vibrant\gulpfile.coffee
[11:51:48] Starting 'coffee'...
[11:51:48] Finished 'coffee' after 11 ms
[11:51:48] Starting 'benchmark'...
[11:51:48] Finished 'benchmark' after 4.42 ms
[11:51:48] Running suite node-vibrant quantizer benchmark [F:\dev\js\node-vibrant\test\vibrant.quantizer
e]...
[11:51:55]    Quantizer: Baseline x 4.25 ops/sec ±6.52% (25 runs sampled)
[11:52:01]    Quantizer: MMCQ x 16.95 ops/sec ±4.78% (42 runs sampled)
[11:52:01] Fastest test is Quantizer: MMCQ at 4.0x faster than Quantizer: Baseline
```
32bit pixel manipulation does not contribute much to the performance.
